src/main/kotlin/com/zoltu/gradle/plugin/GitVersioning.kt

```
@@ -64,25 +72,59 @@ class GitVersioning : Plugin<Project> {
 			}
 		}
 		(project.tasks ?: throw Exception("The project has no tasks.")).withType(Jar::class.java, MethodClosure(closure, "apply"))
 	}
 
-	private fun exposeVersionInfoToUser(project: Project, versionInfo: VersionInfo) {
-		project.extensions.create("ZoltuGitVersioning", Extension::class.java, versionInfo)
-	}
+	private fun exposeConfigurationObject(project: Project, configurationChangeCallback: () -> Unit) = project.extensions.create("ZoltuGitVersioning", Configuration::class.java, configurationChangeCallback)
 
-	open class Extension(open val versionInfo: VersionInfo)
+	open class Configuration(private val configurationChangeCallback: () -> Unit) {
+		companion object {
+			@Suppress("unused")
+			const val NAME = "versioning"
+			private val defaultVersionToString = Function<VersionInfo, String> { versionInfo: VersionInfo ->
+				val suffix = if (versionInfo.tags != null && versionInfo.patch != null) {
+					// semantic versioning
+					"${versionInfo.patch}-${versionInfo.tags}.${versionInfo.commitCount}"
+				} else if (versionInfo.patch != null) {
+					// semantic versioning
+					"${versionInfo.patch}-${versionInfo.commitCount}"
+				} else if (versionInfo.tags != null) {
+					// simple versioning
+					"${versionInfo.commitCount}-${versionInfo.tags}"
+				} else {
+					// simple versioning
+					"${versionInfo.commitCount}"
+				}
+				"${versionInfo.major}.${versionInfo.minor}.$suffix".toString()
+			}
+			private val defaultCustomDescribeProcessor = Function<String, VersionInfo?> { null }
+		}
+		/**
+		 * Contains version information for consumption.
+		 */
+		lateinit var versionInfo: VersionInfo
+
+		/**
+		 * A custom describe processor that will take in a string as input and return a VersionInfo object as output.
+		 */
+		var customDescribeProcessor: Function<String, VersionInfo?> by observable(defaultCustomDescribeProcessor, { property, old, new -> configurationChangeCallback() })
```

I have some concerns here:

Instantiating a new instance of `VersionInfo` in a `build.gradle` script is very clunky because of Groovy's dynamic typing. For example, Intellij IDEA will be unable to resolve the `com.zoltu.gradle.plugin.VersionInfo` symbol and will display an error, even though the code compiles. It's possible to work around this issue by including the plugin as a dependency, but I think it would be better to ask the user for a regex instead of a factory.

---

I actually started with a regex provider, as you had.  The problem there was that it required the user to have a regex with the right number of captures in the right places.  As I was writing the doc-comment and error handling for it, I realized that it was complex to explain to end users and *really* prone to errors, which is why I fell-back to the factory mechanism instead.

I wonder if there is some way to get strong typing on the `VersionInfo` object in IntelliJ, ideally with auto-complete?  I wonder if Kotlin gradle files have the same problem or not?

Another option would be to have the factory return an array of Strings?  I don't like this because instead of failing fast, you fail slow and debugging is more difficult but presumably logic could be added in the plugin to give very clear messages here.  This will "solve" the IntelliJ problem, but at a cost.

Another option would be to use a RegEx with named captures rather than ordered captures.  I haven't tried this in Java, is it a thing one can do (my RegEx-foo is weak)?

---

> I wonder if there is some way to get strong typing on the VersionInfo object in IntelliJ, ideally with auto-complete? I wonder if Kotlin gradle files have the same problem or not?

The easiest workaround is just adding a dependency on the plugin in the usual place; that will allow IDEA to resolve the type and remove the error. We could even do this during plugin application to make the process transparent. But asking the user to add another dependency just for this plugin is a really non-ideal solution.

> Another option would be to use a RegEx with named captures rather than ordered captures. I haven't tried this in Java, is it a thing one can do (my RegEx-foo is weak)?

So is mine :sob:

According to [this](http://stackoverflow.com/a/415635/2286903) answer Java has supported named capture groups since Java 7. I'm not sure how this would make a regex provider more error-resistant, though? It seems like it would be easiest to just kindly suggest to the user that they not specify a custom regex provider if they are not 100% confident in what they are doing.

I understand where you're coming from with regards to how difficult it is to safely write or edit a complex regular expression, but I feel like the negatives of the factory-based approach outweigh the negatives of a regex provider.

---

Named captures would make it so the user's desired string didn't have to match the expected order.  For example, they could have `version-patch.minor.major` as their tagging mechanism.  If we just ask for an arbitrary RegEx, then the user is forced into a regex that matches the expected pattern.

Out of curiosity, what is the use case that you are looking to provide for?  Presumably you created this PR to allow you to solve a particular problem, maybe if I understood that it would help understand the best solution for you?

In hindsight, I think allowing for custom expression/factory is what I should have done in the first place, rather than adding support for 2 more variations on the default.  If we move forward with this change, I'll probably do a major version bump and remove the other two in favor of a single default with customization.

---

> adding a dependency on the plugin in the usual place

What is the "usual place" you are referring to?  You mean in the `buildscript` block?  Is it possible for a plugin to expose a class to the rest of Gradle such that IntelliJ picks it up?  If so, that seems like the best option.

---

> Out of curiosity, what is the use case that you are looking to provide for? Presumably you created this PR to allow you to solve a particular problem, maybe if I understood that it would help understand the best solution for you?

I originally made this PR because I wanted to use a custom implementation of `VersionInfo.toString()` and I didn't realize until later that I could have just done something like:

```groovy
version "${ZoltuGitVersioning.versionInfo.major}.${ZoltuGitVersioning.versionInfo.minor}" +
        ".${ZoltuGitVersioning.versionInfo.patch}.${ZoltuGitVersioning.versionInfo.commitCount}"
```

Oh well :grimacing:

> What is the "usual place" you are referring to? You mean in the buildscript block? Is it possible for a plugin to expose a class to the rest of Gradle such that IntelliJ picks it up? If so, that seems like the best option.

```groovy
repositories {
    maven {
        url 'https://plugins.gradle.org/m2/'
    }
}

dependencies {
    compile group: 'gradle.plugin.com.zoltu.gradle.plugin',
            name: 'git-versioning',
            version: '2.0.34'
}
```

But this does mean that we're building an unnecessary extra dependency, which some users might not like.

---

Hmm, I am surprised that adding a compile time dependency would influence the gradle script auto-completion engine in IntelliJ?  If that is how IntelliJ works then that sounds like a bug to me.  IntelliJ should be using resolved `plugins { ... }`  and `buildscript { ... }` stuff for auto-completion and inline checks.

I'm coming around to the idea of allowing for a named-capture regex.  Maybe we can provide both the factory option and the named capture as extensibility points, and then I can see if I can figure out a "good" way to expose the `VersionInfo` object to gradle, and can also see how it works when you have a Kotlin gradle file.  I'll think on it for a bit and then make a decision.

---

A little more research lead me to find that named captures are not available in Kotlin until 1.1, due to the current tight coupling with JDK 6 which doesn't support named captures: https://youtrack.jetbrains.com/issue/KT-12753

I attempted to switch the describe processor from returning `VersionInfo` to returning a map only to realize that the way things are setup now is to fail fast, so if the describe output doesn't initially start with something that matches one of the existing patterns then the plugin will fail before the user gets a chance to apply configuration.  The correct way to resolve this is to not apply versioning until *after* the gradle file has been parsed.  However, if I make that change then the `VersionInfo` data won't be available until after the gradle file has been parsed, which means you wouldn't be able to do `version "${ZoltuGitVersioning.versionInfo.major}..." in your `build.gradle` file.

I could switch it to silently fail, but that leads to user confusion when their project isn't setup right.  I could continue on with not evaluating the versioning until after `build.gradle` has been evaluated and require users to set whatever they want in a closure, but that leads to `build.gradle` complexity for simple scenarios.

I'm going to think on this some more over-night.  At the least, we can get the change you wanted in, which would allow users to specify a custom `toString()` method.  Unlike your workaround, this would make it so the version printed into the JAR manifest would also match your desired structure, and `./gradlew version` would print the desired version.